src/interactions.js

```diff
@@ -2,98 +2,118 @@
+
+    var interactions = {
+        buttonDown: false,
+        shiftDown: false,
+        lastX: 0.0,
+        lastY: 0.0
+    };
 
     function mousedownFn(e) {
-        if(e.button == 0) {
-            let tmp_interactions = component.state.interactions;
-            tmp_interactions.buttonDown = true;
-            tmp_interactions.lastX = e.clientX;
-            tmp_interactions.lastY = e.clientY;
-
-            component.setState({
-                interactions: tmp_interactions,
-                refreshView: true
-            });
+        if(e.button === 0) {
+            interactions = {
+                buttonDown: true,
+                shiftDown: interactions.shiftDown,
+                lastX: e.clientX,
+                lastY: e.clientY
+            };
         }
     }
     container.addEventListener("mousedown", mousedownFn);
 
     function mouseupFn(e) {
-        if(e.button == 0) {
-            let tmp_interactions = component.state.interactions;
-            if(!tmp_interactions.buttonDown) {
+        if(e.button === 0) {
+            if(!interactions.buttonDown) {
                 return;
             }
 
-            tmp_interactions.buttonDown = false;
-
-             component.setState({
-                interactions: tmp_interactions,
-                refreshView: false
-            });
+            interactions.buttonDown = false;
         }
     }
     window.addEventListener("mouseup", mouseupFn);
 
+    function keychangeFn(e) {
+        interactions.shiftDown = e.shiftKey;
+    }
+    window.addEventListener("keydown", keychangeFn);
+    window.addEventListener("keyup", keychangeFn);
+
     function mousemoveFn(e) {
-        var tmp_interactions = component.state.interactions;
-        if(!tmp_interactions.buttonDown){
+        if(!interactions.buttonDown || (e.buttons === 0)){
             return;
         }
 
-        var dx = e.clientX - tmp_interactions.lastX;
-        var dy = e.clientY - tmp_interactions.lastY;
-        if(dx == 0 && dy == 0) {
+        // prevents interaction with other page elements while dragging
+        e.preventDefault();
+
+        var dx = e.clientX - interactions.lastX;
+        var dy = e.clientY - interactions.lastY;
+        if(dx === 0 && dy === 0) {
             return;
         }
 
-        tmp_interactions.lastX = e.clientX;
-        tmp_interactions.lastY = e.clientY;
-
-        var view = Object.assign({}, component.props.view);
-        speckView.rotate(view, dx, dy);
+        interactions.lastX = e.clientX;
+        interactions.lastY = e.clientY;
 
-        component.props.setProps({
-            view: view
-        });
+        if(interactions.shiftDown) {
+            var translation = getTranslation();
+            var inverseZoom = 0.001/getZoom();
+            setTranslation({
+                x: translation.x - dx * inverseZoom,
+                y: translation.y + dy * inverseZoom
+            });
+        }
+        else {
+            var viewDummyObj = {
```

Comment:
Please also change all `var` to `let` or `const`!

Reply:
If we change all the `var` to `let` or `const`, then we must either bring in a transpiler like babel to convert the code to `ES5` (this is not present at the moment since the entire codebase is `ES5` so this would be redundant) or accept that browser compatibility will decrease as result (https://caniuse.com/#search=let). I thought that since the entire codebase is written in `ES5` I might as well stick with it since the drawbacks of doing so are minimal, but I can switch to `let` or `const` just let me know if either of these two compromises listed above are acceptable.

Reply:
We can stick to ES5 for now, but it might be a good idea in the future to update the codebase. Thank you for clarifying!